hail/Makefile

```diff
@@ -41,13 +42,25 @@ SCALA_BUILD_INFO := src/main/resources/build-info.properties
 
 SHADOW_JAR := build/libs/hail-all-spark.jar
 SHADOW_TEST_JAR := build/libs/hail-all-spark-test.jar
+DEBUG_CLASSES := build/debug
 PYTHON_JAR := python/hail/backend/hail-all-spark.jar
 WHEEL := build/deploy/dist/hail-$(HAIL_PIP_VERSION)-py3-none-any.whl
 EGG := build/deploy/dist/hail-$(HAIL_PIP_VERSION)-py3.6.egg
 
 .PHONY: shadowJar
+ifdef HAIL_DEBUG_MODE
+shadowJar: debugShadowJar
+endif
 shadowJar: $(SHADOW_JAR)
 
+.PHONY: debugShadowJar
+debugShadowJar: $(SHADOW_JAR)
+	mkdir -p build/debug/scala/is/hail/annotations/ && \
+	javac src/debug/scala/is/hail/annotations/Memory.java && \
+	mv src/debug/scala/is/hail/annotations/Memory.class build/debug/scala/is/hail/annotations && \
```

Comment:
Have a look at the `-d` option of `javac`. It does what you want here.

The syntax would be:
```
javac -d build/classes/debug/scala src/debug/scala/is/hail/annotations/Memory.java
```